### PR TITLE
Fixes #30459 - Global Registration Template - Smart Proxy callback support

### PR DESCRIPTION
**Pull request by stejskalleos**
Pass `:url` to the template when rendering via proxy.

**Requirements:**
* Proxy: https://github.com/theforeman/smart-proxy/pull/768
* Katello: https://github.com/Katello/katello/pull/8938

*How to test:*
```
curl proxy.example.com/register | bash
```

### CODE REVIEW

File: app/views/unattended/provisioning_templates/registration/global_registration.erb
Comment (ekohl) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb]: If you pass in a URL that includes get parameters, things break. Perhaps it needs to be a macro that accepts an optional UUID param?

File: app/views/unattended/provisioning_templates/registration/global_registration.erb
Comment (ekohl) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb]: Is this allowed syntax in safemode?
```suggestion
    curl --output $CONSUMER_RPM <%= subscription_manager_configuration_url(hostname: @url_host) %>
```

Comment (stejskalleos) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: But:
```ruby
def test(one = nil, two = true, three = 'something')
  puts one
  puts two
  puts three
end

# Run test()
nil
true
something

# Run test(three: 'text')
{:three=>"text"}
true
something
```

File: app/views/unattended/provisioning_templates/registration/global_registration.erb
Comment (stejskalleos) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: Yeah, but we never do that, see: 
- https://github.com/theforeman/smart-proxy/pull/768/files#diff-45f33bea51f0f06fd02907550e854ae2R26
- https://github.com/theforeman/foreman/pull/8034/files#diff-c7642c50c0b5716cfc364d7e3d0b3589R49

File: app/views/unattended/provisioning_templates/registration/global_registration.erb
Comment (ekohl) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: Ah yes, you need to make it callable like that on the method
```ruby
def test(one = nil, two = true, three: 'something')
  puts one
  puts two
  puts three
end

# Run test()
nil
true
something

# Run test(three: 'text')
nil
true
text
```

File: app/views/unattended/provisioning_templates/registration/global_registration.erb
Comment (ekohl) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: My problem is that building URLs in the template is a bad pattern and should never be encouraged. We did that in the beginning with `foreman_url`. Then at some point we did need a static parameter. In 170b56c69ab0d7e6b332b352d823bb988eca0b6b we added the parameters option to `foreman_url`. In f7e8785ba53f793ca1ab847bdeb1fa07b77da1d7 we synced the use of it.

I think it may actually make sense to extend `foreman_url` so you would end up using something like `foreman_url('register')`. Then for the user it's a consistent API. We also use the proper Rails routing framework rather than hardcoding routes.

A huge benefit of that is that you can refactor things and user templates do not need to be modified. A user may clone the template to make modifications. At that point upgrading becomes painful. If you use a macro, at least that pain disappears.

In short: it's future proofing your code. Future users and developers will be thankful.

Comment (ares) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: We can't use macro here if UUID is only found during the runtime right? Macro is evaluated on the Foreman server before the script gets executed. I agree it's a good practice to avoid this, but how can we achieve it here?

Comment (ekohl) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: You can use the literal `$UUID` as value and rely on shell interpolation (as it does now).

Comment (stejskalleos) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: > You can use the literal $UUID as value and rely on shell interpolation (as it does now).

Sorry I don't understand what does it mean. Can you explain it please?

**Comment by ekohl** [comment]
Slight nit: Capsule is a downstream name. Upstream we use Smart Proxy.

File: app/views/unattended/provisioning_templates/registration/global_registration.erb
Comment (ares) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: I guess that means doing something like this

```ruby
register_path :uuid => '$UUID'
```

to render the path, however

1) that would only be applicable to Katello flow, we'd need that as a separate variable, not based on `@registration_url`
2) a bigger problem is, I don't know how to avoid escaping the `$` sign, rails helpers always do it if I'm not mistaken, we'd need to `CGI::unescape(` that part, but that's very ugly. Or use another placeholder that would be replaced with `$UUID` before rendered to the template. Still ugly, but doable, I'll add the code in the controller comment so we can see how it would look like.

I'd prefer to keep appending `?uuid=$UUID`, the generic register url shouldn't really contain parameters. I have another suggestion though for the controller and how we render the URL there.

File: app/controllers/concerns/foreman/controller/registration.rb
Comment (ares) [inline app/controllers/concerns/foreman/controller/registration.rb]: you can use 

```ruby
register_url :host => url.host
```

to avoid appending `/register`, just in case it ever changes

similarilly you could set the `:protocol => 'https'` or based on `url.scheme`.

You can also define a new variable, registration_url_with_uuid like this

```ruby
registration_url_with_uuid = register_url(:host => url.host, :protocol => url.scheme, :uuid => 'UUID').sub('UUID', '$UUID')
```

that would avoid also appending `?uuid=xyz` in the template itself

**Review by ares — CHANGES_REQUESTED** [review]
Makes sense to me, please:

1) take a look at url creation, with the example from above, we should be able to avoid appending completely

2) rename the Capsule to Proxy in the commit title

3) provide an explaining comment to the commit, namely talking about why the url parameter is added, what it defaults to if it's not set

After that, I think this can be merged

File: app/views/unattended/provisioning_templates/registration/global_registration.erb
Comment (ekohl) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: Perhaps the `register_path` macro should just append `?uuid=$UUID` if rails helpers don't work well for our user case? At least abstract it away from the user?

File: app/controllers/concerns/foreman/controller/registration.rb
Comment (ekohl) [inline app/controllers/concerns/foreman/controller/registration.rb]: We still have support for `RAILS_RELATIVE_URL_ROOT`. If that's used, the Proxy URL would break.

Also note that you don't take the port into consideration so you would lose that info.

Please write some test for the use case where a param is passed in.

**Review by ekohl — DISMISSED** [review]

File: app/controllers/concerns/foreman/controller/registration.rb
Comment (ares) [reply app/controllers/concerns/foreman/controller/registration.rb]: EDIT

Do we really support that? I didn't find a mention in our manual or an installer options. Looking at the puppet-foreman, it seems to be based on `foreman_url` parameter, however it's documentation does not mention such behavior.

~Anyway, the `register_url` respects that variable, link any other helper. E.g.~

The problem is, in such configratuion, the prefix *will be added* even for proxy url, I didn't find a way to ignore such prefix yet

```
RAILS_RELATIVE_URL_ROOT=/foreman be rails c
[1] pry(main)> app.register_path
=> "/foreman/register"
[2] pry(main)> app.register_url
=> "http://www.example.com/foreman/register"
[3] pry(main)> app.register_url host: 'example.tst'
=> "http://example.tst/foreman/register"
```

I agree we need to add port as well. Test would be great for sure.

Comment (ares) [reply app/controllers/concerns/foreman/controller/registration.rb]: Ok, finally found undocummented parameter you can use

```
register_url host: 'example.tst', script_name: ''
=> "http://example.tst/register"
```

so I think required changes are:
1) use empty script_name in case of the url is coming from the parameter *sigh*
2) also pass port parsed out of the url
3) add a test for generating the right values to the variables in all these cases

Thanks for bearing with us.

Comment (ekohl) [reply app/controllers/concerns/foreman/controller/registration.rb]: > Do we really support that? I didn't find a mention in our manual or an installer options. Looking at the puppet-foreman, it seems to be based on `foreman_url` parameter, however it's documentation does not mention such behavior.

It's defined here:
https://github.com/theforeman/foreman/blob/93dd5ad7657023873c87c7aabeba4e9ed23f713e/config.ru#L9

Then it can be set here by the installer:
https://github.com/theforeman/puppet-foreman/blob/495727bdfe4a60df28c542cf5b771a91e20433b3/templates/_suburi.conf.erb#L1

I can't say for sure it still works with Puma out of the box.

However, the UI has repeatedly broken this and I haven't seen anyone complain. Perhaps we should open an RFC to remove it?

> The problem is, in such configratuion, the prefix _will be added_ even for proxy url, I didn't find a way to ignore such prefix yet

That was precisely my point that I should have been explicit about: the Foreman proxy URL should not be based on the Rails URL helper because they're not really related. I still am strongly in favor of using proper URL helpers rather than hardcoding, but with a remote API that's hard to avoid.

I think here it should be:
```ruby
if params[:url].presence
  uri = URI(params[:url])
  uri.path += '/register'
  registration_url = uri.to_s
  uri.query = 'uuid=$UUID'
  registration_url_with_uuid = uri.to_s
else
  registration_url = register_url
  registration_url_with_uuid = register_url(uuid: 'UUID').sub('UUID', '$UUID')
end
```

Comment (stejskalleos) [reply app/controllers/concerns/foreman/controller/registration.rb]: How about `Setting[:foreman_url]`, isn't that more important than `ENV['RAILS_RELATIVE_URL_ROOT']`?
What if they don't have same value, what should be expected behavior?

**Comment by stejskalleos** [comment]
Updated:
* Added tests (I hope I covered all scenarios)
* Using using `register_url` instead of `Setting[:foreman_url]` solved the issue with `ENV['RAILS_RELATIVE_URL_ROOT']`
* Removed `registration_url_with_uuid`, using `<%= @registration_url %>?uuid=$UUID` in template is cleaner than doing the hack `register_url(host: url.host, protocol: url.scheme, uuid: 'UUID').sub('UUID', '$UUID')`

File: app/controllers/concerns/foreman/controller/registration.rb
Comment (ares) [inline app/controllers/concerns/foreman/controller/registration.rb]: s/presence/present?/ to be more semantical, `presence` returns the value we don't use anywhere, `present?` returns boolean, which we're after in this condition

File: app/controllers/concerns/foreman/controller/registration.rb
Comment (ares) [inline app/controllers/concerns/foreman/controller/registration.rb]: please use `if` when the condition has either `||` or `&&`, it's much easier to read.

```
if !uri.kind_of?(URI::HTTP) && !uri.kind_of?(URI::HTTPS)
```

you could also do `if uri.scheme.start_with?('http')` or more explicit form `%w[http https].include?(uri.scheme)` which I personally find more readable and straightforward

File: test/controllers/api/v2/registration_controller_test.rb
Comment (ares) [inline test/controllers/api/v2/registration_controller_test.rb:118]: we should reset the ENV variable to previous value at the end of the test, this could cause weird problems in other tests I think

File: test/controllers/api/v2/registration_controller_test.rb
Comment (ares) [inline test/controllers/api/v2/registration_controller_test.rb:97]: one more such test for `http` remaining `http` please

File: app/views/unattended/provisioning_templates/registration/global_registration.erb
Comment (ares) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb:55]: I see we're still appending `?uuid=$UUID` here, this PR does not make it worse so I'm fine with that. We could play with URI#query, but let's keep that to when we really need it.

**Review by ares — CHANGES_REQUESTED** [review]
few more nits but overall it has the right direction, if suggestions are implemented, I plan to merge this next week (provided testing will be successful)

File: app/controllers/concerns/foreman/controller/registration.rb
Comment (ares) [reply app/controllers/concerns/foreman/controller/registration.rb]: I know it's in the code. As I said, the installer only does that if it sees the suburi in [forman-url](https://github.com/theforeman/puppet-foreman/blob/6f5b99bb08f19ff61410dc6208346b9c8c333221/manifests/config/apache.pp#L137-L152)  however foreman-url documentation [does not mention](https://github.com/theforeman/puppet-foreman/blob/6f5b99bb08f19ff61410dc6208346b9c8c333221/manifests/init.pp#L43) that at all. 

> However, the UI has repeatedly broken this and I haven't seen anyone complain. Perhaps we should open an RFC to remove it?

Yes! Based on today status, I don't think people know about that and I'd also say it's possible but not supported.

> That was precisely my point that I should have been explicit about: the Foreman proxy URL should not be based on the Rails URL helper because they're not really related. I still am strongly in favor of using proper URL helpers rather than hardcoding, but with a remote API that's hard to avoid.

Yeah, got it a second after I sent the first post, then I editted it. I think you are right with your example, however the query part is not that easy. While it would avoid the escaping issue, it overrides any query already present. We'd need to split it by `&`, add a new to the array and the join again and set it only then. In other words, it has the same flaw as just appending, we only get rid of the `?` for the price of more complex code and two different paths handling it. I don't think it's worth it.

Comment (ares) [reply app/controllers/concerns/foreman/controller/registration.rb]: > How about Setting[:foreman_url], isn't that more important than ENV['RAILS_RELATIVE_URL_ROOT']?
What if they don't have same value, what should be expected behavior?

The installer I believe takes care of their values. If user changes that manually, that's a problem. I think in this case, we should either use `Setting[:foreman_url]` as a primary source. If there's external URL, we should ignore the ENV variable, since that's only for internal URLs.

File: app/controllers/concerns/foreman/controller/registration.rb
Comment (ekohl) [inline app/controllers/concerns/foreman/controller/registration.rb:79]: You're not likely to hit this, but this does ensure that the Smart Proxy must be mounted on `/`:
```
[1] pry(main)> require 'uri'
=> false
[2] pry(main)> URI.join('https://example.com/bla', '/register')
=> #<URI::HTTPS https://example.com/register>
```

Is it intention that you ignore the path provided in the URL? On the plus side, it also throws away any GET parameters so it's guaranteed that a string concat later succeeds.

File: app/controllers/concerns/foreman/controller/registration.rb
Comment (ekohl) [inline app/controllers/concerns/foreman/controller/registration.rb]: Does this work? Is `starts_with?` a Rails extension to String? Does it make sense to use the explicit properties on the URI object?
```suggestion
    if uri.scheme && uri.host
```

File: app/controllers/concerns/foreman/controller/registration.rb
Comment (ekohl) [inline app/controllers/concerns/foreman/controller/registration.rb]:
```suggestion
      msg = N_('URL in :url parameter is missing a scheme, please set http:// or https://')
```

File: app/controllers/concerns/foreman/controller/registration.rb
Comment (stejskalleos) [reply app/controllers/concerns/foreman/controller/registration.rb:79]: > Is it intention that you ignore the path provided in the URL

Yes, it is. I wanted to be 100% sure that url will be `.../register`, never came up to my mind that proxy can be mounted somewhere else than `/`. Not sure what to do now.

- If I won't reset path, user can by accident pass URL with path, which will cause errors later in the template.
- But if I will keep the behavior as it is now, URL from proxy mounted somewhere else than `/` simply won't work.

Comment (ekohl) [reply app/controllers/concerns/foreman/controller/registration.rb:79]: I think it's ok to accept it for now. The Proxy isn't mounted elsewhere than `/` in practice so I think it's safe to assume a lot of other things would break if you tried to do so.

**Review by ares — CHANGES_REQUESTED** [review]
During the testing I found out that the final "call home" does not work reliably. The `foreman_url('built')` returns proxy url with ssl, however that does not work if the ssl cert is not installed.

In such case, we also need to deploy the ssl certificate and tell curl to use it.

The relevant code from global template 
```ruby
SSL_CA_CERT=$(mktemp)
cat << EOF > $SSL_CA_CERT
<%= foreman_server_ca_cert %>
EOF
```

This needs to be added to linux_registration_default with also the following logic.

```
curl <%= '--cacert $SSL_CA_CERT' if foreman_url('built').start_with?('https') %> -o /dev/null --noproxy \* '<%= foreman_url('built') %>'
```

You may also want to wrap the first snippet in the same condition. In fact the result of `foreman_url('built').start_with?('https')` should be probably stored in `@deploy_ssl` variable in the controller, to avoid adding more logic to the template.

File: app/controllers/concerns/foreman/controller/registration.rb
Comment (ekohl) [inline app/controllers/concerns/foreman/controller/registration.rb:33]: Not strictly needed and certainly not in this PR, but I feel like this hash is getting long. Perhaps it would be good to spit it up:

```ruby
context = {
  user: User.current,
  # .. other ones
}

params.permit(permitted).to_h.symbolize_keys.merge(context)
```

**Review by ekohl — COMMENTED** [review]
Overall :+1:. Small nit inside that you may want to consider as a follow up. I also know that @ares is testing this so I'll leave it up to him to merge.

**Review by ares — APPROVED** [review]
Thanks @stejskalleos, merging! It works find also for call home callback. The optional refactoring is now tracked as https://projects.theforeman.org/issues/31177 and I'm sending a PR with the suggestion from here.

**Pull request merged**
